### PR TITLE
Implement Sedate IXDTF string parsing

### LINKED ISSUE #2127: Add Sedate IXDTF (extended ISO) string parsing function
We should add a function that parses ISO-8601 strings. We should follow the Temporal grammar:

https://tc39.es/proposal-temporal/#sec-temporal-iso8601grammar

i.e., we should implement a string parser conforming to [IETF Sedate](https://datatracker.ietf.org/wg/sedate/about/) that parses a string into its components (date, time, time zone, calendar), and use this parser in APIs and anywhere else we parse date strings. Think about where to put this code in the crate structure.

We should start with the [DateTime](https://tc39.es/proposal-temporal/#prod-DateTime) production, and work our way up towards [CalendarDateTime](https://tc39.es/proposal-temporal/#prod-CalendarDateTime) and [TemporalZonedDateTimeString](https://tc39.es/proposal-temporal/#prod-TemporalZonedDateTimeString).

Examples of strings that are accepted as part of the DateTime grammar:

- `2022-06-27T16:31:00.000`
- `2022-06-27t16:31:00.000`
- `2022-06-27 16:31:00.000`
- `20220627 16:31:00.000`
- `20220627 163100.000`
- `20220627 163100`
- `20220627 1631`

### PR DESCRIPTION
**Pull request by nekevss**
Hi! This PR is intended to close #2127. 

It would add the `IxdtfParser` and `IsoDurationParser`.

I'm opening this as a draft for a bit early feedback and comments.

This is essentially a proposal to pull in the core of the parser from over in `temporal_rs` along with some updates I've added to allow the parser to function as a standalone as well as some general clean up.

There's still a couple things I have left to do:

  - [x] ALOT more documentation is needed. (More docs added, can be built out further if wanted)
  - [x] Clean up the tests / potentially add more tests.
  
 The one question I have is that the previous experimental version of the parser had a bit more precision on the errors when it came to parsing digits. ~~Currently, this really only throws an `ExpectedDigitChar` error from `Cursor::next_digit`. If that is too general, this could be caught/mapped in the individual functions or `next_digit` could return an `Option<u8>` to allow more exact error handling.~~

EDIT: I did end up removing `ExpectedDigitChar` as an error.

### CODE REVIEW

**Comment by nekevss** [comment]
Marking as ready for review 😄

A few general questions:

- What would be the preferred way of resolving an ambiguity with UTC offsets and Time Zone Annotations? Currently, this defaults to the tz annotation if provided, but maybe this isn't the best approach. It could be better to provide the UTC offset value separate from the tz annotation and allow the caller to handle any ambiguous cases.

- Would it be preferred for `IsoParseRecord` to not have as many nested records as it currently does? (primary cases: `DateRecord` and `TimeRecord`).

- `IxdtfParser` has a lot of parsing options beyond just extended date time (admittedly, a decent portion targeting some Temporal functionality). Would it be preferred to split some of the parsing like time zone and time into a `TemporalParser` over `IxdtfParser`?

File: utils/ixdtf/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+impl ParserError {
+    pub(crate) fn abrupt_end() -> Self {
+        ParserError::AbruptEnd
+    }
+}
+
+impl core::fmt::Display for ParserError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let msg = match self {
+            ParserError::ParseFloat => "Invalid float while parsing fraction part.",
```
Comment (sffc) [inline utils/ixdtf/src/error.rs:70]: Suggestion: use `displaydoc` to put these display strings inline with the enum definitions. We use this dependency widely in icu4x code.

File: utils/ixdtf/src/lib.rs [hunk 2]
```diff
@@ -2,11 +2,109 @@
...
+//! - `2024-03-02T08:48:00-05:00[u-ca=iso8601]`
+//! - `2024-03-02T08:48:00-05:00[u-ca=iso8601][u-ca=japanese]`
+//! - `2024-03-02T08:48:00-05:00[u-ca=iso8601][!answer-to-universe=fortytwo]`
+//!
+//! #### Invalid Key-Value Annotations
+//!
+//! - `2024-03-02T08:48:00-05:00[u-ca=iso8601]`
+//! - `2024-03-02T08:48:00-05:00[u-ca=iso8601][!u-ca=japanese]`
+//! - `2024-03-02T08:48:00-05:00[u-ca=iso8601][!answer-to-universe=fortytwo]`
+//!
```
Comment (sffc) [inline utils/ixdtf/src/lib.rs:96]: Observation: why are the "valid" and "invalid" examples the same?

File: utils/ixdtf/src/parsers/mod.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+// ==== Mini cursor implementation for Iso8601 targets ====
+
+/// `Cursor` is a small cursor implementation for parsing Iso8601 grammar.
+#[derive(Debug)]
+pub(crate) struct Cursor {
+    pos: u32,
+    source: Vec<char>,
+}
```
Comment (sffc) [inline utils/ixdtf/src/parsers/mod.rs:225]: Suggestion: Ideally `Cursor` would borrow the `str`, like

```rust
pub(crate) struct Cursor<'a> {
    pos: u32,
    source: &'a str,
}
```

This would allow the `slice` function to return `&'a str`, like

```rust
impl<'a> Cursor {
    fn slice(&self, start: usize, end: usize) -> Option<&'a str> {
        self.source.get(start..end)
    }
}
```

This is preferable because we reduce memory allocations, which are slow especially in low-level parser code.

The only reason to convert to a `Vec<char>` would be if you need random access. However, I don't _think_ you need this, because

1. I think you are almost always stepping forwards one code point at a time. You can do this in UTF-8 without needing true random access.
2. IXDTF strings should be ASCII, right? So if you ever encounter a non-ASCII code point, you can error out, and then if you do need random access, you can use byte indices anyway.

File: utils/ixdtf/src/parsers/mod.rs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    fn slice(&self, start: u32, end: u32) -> String {
+        let slice = self.source.get(start as usize..end as usize);
+        let Some(slice) = slice else {
+            return String::default();
+        };
+        slice.iter().collect()
+    }
+
+    /// Get current position
+    const fn pos(&self) -> u32 {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/mod.rs:247]: Question: why `u32` instead of `usize`? Seems not necessary to restrict it. These strings should definitely not be that long, but in general it's better for style and security to stick with the native size types. ICU4C has gotten in a little bit of trouble for using `i32` instead of `size_t`, for example.

File: utils/ixdtf/src/parsers/mod.rs [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    }
+
+    /// Parses the source as a [DateTime string][temporal-dt].
+    ///
+    /// This is the baseline parser where the TimeRecord, UTCOffset, and all annotations are optional.
+    ///
+    /// [temporal-dt]: https://tc39.es/proposal-temporal/#prod-TemporalDateTimeString
+    pub fn parse_date_time(&mut self) -> ParserResult<IsoParseRecord> {
+        datetime::parse_annotated_date_time(DateTimeFlags::empty(), &mut self.cursor)
+    }
```
Comment (sffc) [inline utils/ixdtf/src/parsers/mod.rs:68]: Discussion: I think the code that runs the parser differently depending on the Temporal type should stay in the `temporal_rs` crate. It would be reasonable to have flags that control the parser behavior, but how to interpret the return value should be up to the client.

How much do we need these parse flags? I feel like the IXDTF parser can not take any flags and just always return a struct with a bunch of optional fields, plus a map/vec at the end for extra annotations (beyond time zone and calendar).

Basically, the `ixdtf` crate should do one thing only: take a string that is expected to be in IXDTF syntax and parse it into a Rust data structure. Then the consumers of the crate can decide what to do with that structured data.

File: utils/ixdtf/src/parsers/records.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+pub struct IsoParseRecord {
+    /// Parsed Date Record
+    pub date: DateRecord,
+    /// Parsed Time
+    pub time: Option<TimeRecord>,
+    /// Parsed `TimeZone` data (UTCOffset | IANA name)
+    pub tz: Option<TimeZone>,
+    /// The parsed calendar value.
+    pub calendar: Option<String>,
+}
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:23]: Suggestion: as noted elsewhere, I think this should end up borrowing more data, and it should probably have a place to put unknown annotations, like this

```rust
pub struct IsoParseRecord<'a> {
    /// Parsed Date Record
    pub date: DateRecord,
    /// Parsed Time
    pub time: Option<TimeRecord>,
    /// Parsed `TimeZone` data (UTCOffset | IANA name)
    pub tz: Option<TimeZone<'a>>,
    /// The parsed calendar value.
    pub calendar: Option<&'a str>,
    /// Any additional annotations beyond time zone and calendar
    pub annotations: Vec<Annotation<'a>>,
}

pub struct Annotation<'a> {
    pub key: &'a str,
    pub values: &'a str,
    pub critical_flag: bool,
}
```

(it also needs a way to record whether the time zone or calendar had the critical flag)

**Review by sffc — CHANGES_REQUESTED** [review]
Good work! Leaving some architectural feedback first; I noted some of these things when I took a pass over the code in the `temporal_rs` repository, but I didn't post them at that time because they seem like incremental changes from the code base.

File: utils/ixdtf/src/lib.rs (hunk 2 above)
Comment (nekevss) [reply utils/ixdtf/src/lib.rs:96]: Good catch! I missed this. In the second one, it should be due to the critical flag, but the examples are definitely messed up. And honestly, they could probably be better now that I'm looking at it. Grammar could be more firmly laid out as to what characters are expected.

File: utils/ixdtf/src/parsers/mod.rs (hunk 3 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/mod.rs:225]: Yeah, agreed. This is the one improvement that I really think should happen as well. I'll take a look at it.

The one point that is going to be a bit weird as far as Temporal compatibility is the [`TemporalSign`](https://tc39.es/ecma262/#prod-TemporalSign), which isn't completely ASCII.

```txt
[TemporalSign] :::
    ASCIISign
    <MINUS>
```

File: utils/ixdtf/src/parsers/records.rs (hunk 6 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/records.rs:23]: That should be doable.

Also, on the subject of the `IsoParseRecord`, the trend mostly seems the `ixdtf` should just parse the string and provide the record to the user. So should the potential `tz` ambiguity be provided to the user as well.

```rust
pub struct IsoParseRecord<'a> {
    /// Parsed Date Record
    pub date: DateRecord,
    /// Parsed Time
    pub time: Option<TimeRecord>,
    /// Parsed `UTCOffset`.
    pub offset: Option<UtcOffset>,
    /// Parsed `TimeZone` data (UTCOffset | IANA name)
    pub tz: Option<TimeZone<'a>>,
    /// The parsed calendar value.
    pub calendar: Option<&'a str>,
    /// Any additional annotations beyond time zone and calendar
    pub annotations: Vec<Annotation<'a>>,
}
```

Comment (sffc) [reply utils/ixdtf/src/parsers/records.rs:23]: Yes, as far as `ixdtf` is concerned, returning `offset` and `tz` as separate fields sounds like a great idea.

File: utils/ixdtf/src/parsers/mod.rs (hunk 3 above)
Comment (sffc) [reply utils/ixdtf/src/parsers/mod.rs:225]: Acknowledged that there might be a non-ASCII minus sign. So that invalidates point 2, but I think point 1 should still apply. You are pretty much only ever walking forwards, so you shouldn't ever need to do random access. Taking slices on byte indices should be fine.

**Comment by nekevss** [comment]
General updates were made concerning the above review. There are a couple more points that need to still be addressed, but I had a question that I thought I'd pose early. 

Is there any preference when handling parsing MonthDay, YearMonth, and annotated Time strings? 

I'm tempted to keep those but include them as part of a separate parser from the current `IxdtfParser`, but I thought I'd float the question in case there was any feedback regarding go forward on those.

File: utils/ixdtf/src/parsers/datetime.rs [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    let first = cursor.next_digit()?.ok_or(ParserError::DateDay)?;
+    let day_value = first * 10 + cursor.next_digit()?.ok_or(ParserError::DateDay)?;
+    Ok(day_value)
+}
+
+#[inline]
+fn check_date_validity(year: i32, month: u8, day: u8) -> ParserResult<()> {
+    if !(-271_820..=275_760).contains(&year) {
+        return Err(ParserError::InvalidYearRange);
+    }
```
Comment (sffc) [inline utils/ixdtf/src/parsers/datetime.rs:244]: Question: is this an IXDTF limitation or a Temporal limitation?

File: utils/ixdtf/src/parsers/time.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    };
+    Ok(to_fraction_components(fraction))
+}
+
+/// Parse a `Fraction` value
+///
+/// This is primarily used in ISO8601 to add percision past
+/// a second.
+#[inline]
+pub(crate) fn parse_fraction(cursor: &mut Cursor) -> ParserResult<Option<f64>> {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/time.rs:178]: Suggestion: I think `f64` is not the right choice here since it might be lossy; a `u32` of nanoseconds should be a better choice.

File: utils/ixdtf/src/parsers/time.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    // Assert that the first char provided is a decimal separator.
+    if !cursor.check_or(false, is_decimal_separator) {
+        return Ok(None);
+    }
+    cursor.next_or(ParserError::FractionPart)?;
+    let mut fraction = String::from('.');
```
Comment (sffc) [inline utils/ixdtf/src/parsers/time.rs:184]: Issue: Try to avoid internal allocations like this while parsing.

File: utils/ixdtf/src/parsers/time_zone.rs [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+/// [tz]: https://tc39.es/proposal-temporal/#prod-TimeZoneIdentifier
+pub(crate) fn parse_time_zone<'a>(cursor: &mut Cursor<'a>) -> ParserResult<TimeZone<'a>> {
+    let is_iana = cursor
+        .check(is_tz_leading_char)
+        .ok_or(ParserError::abrupt_end("TimeZoneAnnotation"))?;
+    let is_offset = cursor.check_or(false, is_sign);
+
+    if is_iana {
+        return parse_tz_iana_name(cursor);
+    } else if is_offset {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/time_zone.rs:112]: Question: is this code path for something like `[-08:00]`? Does the IXDTF syntax allow that?

File: utils/ixdtf/src/parsers/records.rs [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+pub struct Annotation<'a> {
+    pub critical: bool,
+    pub key: &'a str,
+    pub value: &'a str,
+}
+
+#[non_exhaustive]
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:37]: Suggestion: `DateRecord` and `TimeRecord` can probably be made exhaustive since by specification these are the only fields that could appear. Probably good to keep `IsoParseRecord` as non-exhaustive since in principle we could pull out more of the annotations to fields. Not sure about `Annotation`.

File: utils/ixdtf/src/parsers/records.rs [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+
+/// `TimeZone` data
+#[non_exhaustive]
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct TimeZone<'a> {
+    /// TimeZoneIANAName
+    pub name: Option<&'a str>,
+    /// TimeZoneOffset
+    pub offset: Option<UTCOffset>,
+}
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:75]: Suggestion: can you make this an enum between the two types? Seems like a bug if this ever returns `Some/Some` or `None/None`

File: utils/ixdtf/src/parsers/records.rs [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    /// A minute value
+    pub minute: u8,
+    /// A second value.
+    pub second: u8,
+    /// A millisecond value
+    pub millisecond: u16,
+    /// A Microsecond value
+    pub microsecond: u16,
+    /// A Nanosecond value
+    pub nanosecond: u16,
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:64]: Suggestion (optional), here and below: I probably lean toward a single `nanosecond: u32` since (1) it is fewer bits overall, (2) there are fewer invariants to check, and (3) it is easier to reconstruct a `second` value.

File: utils/ixdtf/src/parsers/records.rs [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    /// The `minutes` value.
+    pub minutes: f64,
+    /// The `seconds` value.
+    pub seconds: f64,
+    /// The `milliseconds` value.
+    pub milliseconds: f64,
+    /// The `microseconds` value.
+    pub microseconds: f64,
+    /// The `nanoseconds` value.
+    pub nanoseconds: f64,
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:121]: Discussionm, here and below: it looks like you made these public fields `f64` since they could be either integral or fractional in the ISO Duration syntax. Is that right? Are there cases where the `f64` could lose precision? Did you consider other ways of expressing these quantities?

**Review by sffc — COMMENTED** [review]
Nice work! I did a more thorough review this time. Just some questions mostly about data types. Feel free to convert comments into follow-up issues as you see fit. I would like to land this soon and favor follow-up PRs.

File: utils/ixdtf/src/parsers/mod.rs [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+    pub fn parse(&mut self) -> ParserResult<IsoParseRecord> {
+        datetime::parse_annotated_date_time(&mut self.cursor)
+    }
+}
+
+/// An extended parser for parsing additional extended datetime formats like
+/// `AnnotatedYearMonth`s, `AnnotatedMonthDay`s and `AnnotatedTime`s
+#[cfg(feature = "temporal")]
+#[derive(Debug)]
+pub struct TemporalParser<'a> {
```
[the remark this answers is not in the document]
Comment (nekevss) [reply utils/ixdtf/src/parsers/mod.rs:75]: Yeah, the separation primarily came from MonthDay and Time needing special handling. I'd be fine with adding an option onto `IxdtfParser`.

File: utils/ixdtf/src/parsers/time_zone.rs (hunk 10 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/time_zone.rs:112]: I'm fairly certain it does in both Temporal and Sedate.

Sedate lists a time zone annotation as below

```txt
Sedate:
time-zone         = "[" critical-flag
                        time-zone-name / time-numoffset "]"

RFC3339 5.6:
time-numoffset  = ("+" / "-") time-hour ":" time-minute
```

File: utils/ixdtf/src/parsers/records.rs [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+//! Parse records for `ixdtf`'s parsers to target.
+
+use alloc::vec::Vec;
+
+/// An `IsoParseRecord` is an intermediary record returned by ISO parsing functions.
+///
+/// `IsoParseRecord` is converted into the ISO AST Nodes.
+#[non_exhaustive]
+#[derive(Default, Debug, PartialEq)]
+pub struct IsoParseRecord<'a> {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:14]:
```suggestion
pub struct IxdtfParseRecord<'a> {
```

File: utils/ixdtf/src/parsers/records.rs [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    /// A Microsecond value
+    pub microsecond: u16,
+    /// A Nanosecond value
+    pub nanosecond: u16,
+}
+
+/// `TimeZone` data
+#[non_exhaustive]
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct TimeZone<'a> {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:70]: Suggestion: most of the structs in this file say `Record`. Probably be consistent here too with `TimeZoneRecord`. Also `UtcOffsetRecord`

File: utils/ixdtf/src/parsers/records.rs (hunk 14 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/records.rs:121]: I had partially played around with changing these fields. They were mostly the drivers for keeping the fraction parsing as allocating to a `String` that was parsed into a `f64` over a `u32` approach (P1.005H is different than 10:12:08.000123456). I'll take a look at this again with the fraction change / any refactor.

I'm not entirely sure if there are cases where `f64` loses precision. The fraction part can only be 9 decimal places, at least going according the grammar laid out in Temporal. No matter how much research I did, I wasn't able to find any upper and lower bounds on the duration field values (outside of the restriction on Duration itself).

If you have any ideas and/or suggestions, I'd definitely be open to them.

File: utils/ixdtf/src/parsers/datetime.rs (hunk 7 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/datetime.rs:244]: This limit is from the Temporal specification. It can be removed and enforced on the Temporal side

File: utils/ixdtf/src/parsers/records.rs (hunk 14 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/records.rs:121]: I think there may be an option to simply parse the components and provide them as one record without manually handling the fraction part, which would simplify everything.

```rust

pub struct Duration {
  pub years: i32,
  pub months: i32,
  pub weeks: i32,
  pub days: i32,
  pub hours: Option<i32>,
  pub minutes: Option<i32>,
  pub seconds: Option<i32>,
  pub fraction: Option<(u32, u8)>
}
```

The above would essentially be true where the FractionPart applies to the smallest available unit.

Comment (sffc) [reply utils/ixdtf/src/parsers/records.rs:121]: Yeah I like something along those lines, or maybe instead of having all the `Option`s, use an enum

```rust
pub enum DurationFraction {
    Hours(u32),
    Minutes(u32),
    Seconds(u32),
}
```

File: utils/ixdtf/src/parsers/records.rs (hunk 11 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/records.rs]: I did remove the `#[non_exhaustive]` from the struct. 

As a general side note / follow up, even if the records are exhaustive, would there be any benefit in keeping them as `#[non_exhaustive]` for semver considerations and to also prevent the structs from being constructed by a `StructExpression`? Although, I don't really see any reason any `ixdtf` user would have a need for that functionality.

Comment (sffc) [reply utils/ixdtf/src/parsers/records.rs]: Exhaustive structs are a bit more useful to users because you can unpack them / match on them without needing a catchall code path, but non-exhaustive lets the ICU4X team add things without breaking semver. We have this lint primarily so that we make an explicit choice for each thing we export

File: utils/ixdtf/src/lib.rs [hunk 18]
```diff
@@ -2,11 +2,197 @@
...
+//! let date = result.date.unwrap();
+//! let time = result.time.unwrap();
+//! let tz = result.tz.unwrap();
+//!
+//! assert_eq!(date.year, 2024);
+//! assert_eq!(date.month, 3);
+//! assert_eq!(date.day, 2);
+//! assert_eq!(time.hour, 8);
+//! assert_eq!(time.minute, 48);
+//! assert_eq!(tz.name, Some("America/New_York"));
```
[the remark this answers is not in the document]
Comment (sffc) [reply utils/ixdtf/src/lib.rs]: ^ Please add an assertion for the `-05:00` which is also in the string

File: utils/ixdtf/src/parsers/mod.rs [hunk 19]
```diff
@@ -38,6 +36,16 @@ macro_rules! assert_syntax {
...
 
+/// Parsing options that can be provided to `IxdtfParser` to change parsing behavior.
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum IxdtfOptions {
+    None,
+    YearMonth,
+    MonthDay,
+    Time,
+}
```
Comment (sffc) [inline utils/ixdtf/src/parsers/mod.rs:47]: Nit: Add docs

File: utils/ixdtf/src/parsers/datetime.rs [hunk 20]
```diff
@@ -78,13 +79,24 @@ pub(crate) fn parse_annotated_date_time<'a>(
 }
 
 /// Parses a `DateTime` record.
-fn parse_date_time(cursor: &mut Cursor) -> ParserResult<DateTimeRecord> {
-    let date = parse_date(cursor)?;
+fn parse_date_time(cursor: &mut Cursor, options: IxdtfOptions) -> ParserResult<DateTimeRecord> {
+    if options == IxdtfOptions::Time {
+        let start = cursor.pos;
+        let result = parse_ambiguous_time_record(cursor)?;
+        if let Some(success) = result {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/datetime.rs:86]: Suggestion/Issue: I don't think I like the graceful recovery here. If you pass `IxdtfOptions::MonthDay`, and the first field is not a month, then the parse should fail, and if desired, the caller can try again with a different `IxdtfOptions`. It also makes it harder to reason about the code since these inner functions like `parse_ambiguous_time_record` might do funny things with `cursor`.

File: utils/ixdtf/src/parsers/records.rs [hunk 21]
```diff
@@ -28,13 +29,14 @@ pub struct IsoParseRecord<'a> {
...
 #[derive(Debug, Clone, PartialEq)]
+/// A record of an annotation.
 pub struct Annotation<'a> {
     pub critical: bool,
     pub key: &'a str,
     pub value: &'a str,
 }
 
-#[non_exhaustive]
+#[allow(clippy::exhaustive_structs)]
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:39]: Nit: `#[allow(clippy::exhaustive_structs)]` should have a comment on the same line with a brief justification, such as

```suggestion
#[allow(clippy::exhaustive_structs)] // IXDTF defines Date as having three fields
```

File: utils/ixdtf/src/parsers/records.rs [hunk 22]
```diff
@@ -56,28 +58,24 @@ pub struct TimeRecord {
...
-    pub offset: Option<UTCOffset>,
+    Offset(UTCOffsetRecord),
 }
 
 /// A full precision `UtcOffset`
 #[non_exhaustive]
 #[derive(Debug, Clone, Copy, PartialEq)]
-pub struct UTCOffset {
+pub struct UTCOffsetRecord {
     /// The `+`/`-` sign of this `UtcOffset`
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:79]: Nit: do one of the following

1. Document what the output values are. If the number is positive, is `sign > 0`? `sign >= 0`? `sign == 1`?
2. Switch to using an enum
3. Switch to using a bool

File: utils/ixdtf/src/parsers/records.rs [hunk 23]
```diff
@@ -108,62 +114,36 @@ pub struct DurationParseRecord {
...
-        };
-
-        let microseconds = (milliseconds % 1.0) * 1000.0;
-        let nanoseconds = (microseconds % 1.0) * 1000.0;
-
-        let sign: i32 = if sign { 1 } else { -1 };
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DurationFraction {
+    /// The fraction value applied to an hour.
+    Hours(i64),
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:132]: Question: why is this signed?

Comment (sffc) [reply utils/ixdtf/src/parsers/records.rs:132]: Does ISO-8601 allow mixed signed duration strings? I think the answer is No, in which case you should make all the fields unsigned and introduce a single sign field, similar to UtfOffsetRecord.

Comment (sffc) [reply utils/ixdtf/src/parsers/records.rs:132]: Basically my guiding principle is that these records should encode as many invariants as possible. If it isn't possible for there to be mixed signs, then the record should reflect that in its data types.

File: utils/ixdtf/src/parsers/records.rs (hunk 22 above)
Comment (sffc) [reply utils/ixdtf/src/parsers/records.rs:79]: Assuming you also use this in DurationParseRecord, I think it's probably worth making it an enum.

File: utils/ixdtf/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (nekevss) [reply utils/ixdtf/src/lib.rs]: Sorry about that! For some reason, I thought I had updated this 😕 Made sure to update it

File: utils/ixdtf/src/parsers/datetime.rs [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+        1 | 3 | 5 | 7 | 8 | 10 | 12 => Some(31),
+        4 | 6 | 9 | 11 => Some(30),
+        2 => Some((28 + in_leap_year(year)) as u8),
+        _ => None,
+    }
+}
+
+/// Utility that returns whether a year is a leap year.
+#[inline]
+fn in_leap_year(year: i32) -> i8 {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/datetime.rs:320]: Suggestion: return a `bool` instead of an int?

File: utils/ixdtf/src/parsers/mod.rs [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+        if !$cond {
+            return Err(ParserError::$err);
+        }
+    };
+}
+
+/// Parsing options that can be provided to `IxdtfParser` to change parsing behavior.
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum IxdtfOptions {
```
Comment (sffc) [inline utils/ixdtf/src/parsers/mod.rs:42]: Suggestion: We should bikeshed the name of this enum.

File: utils/ixdtf/src/parsers/mod.rs [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+    /// This is the baseline parser where the TimeRecord, UTCOffset, and all annotations are optional.
+    ///
+    /// [temporal-dt]: https://tc39.es/proposal-temporal/#prod-TemporalDateTimeString
+    pub fn parse(&mut self) -> ParserResult<IxdtfParseRecord<'a>> {
+        match self.options {
+            IxdtfOptions::None => datetime::parse_annotated_date_time(&mut self.cursor),
+            IxdtfOptions::YearMonth => datetime::parse_annotated_year_month(&mut self.cursor),
+            IxdtfOptions::MonthDay => datetime::parse_annotated_month_day(&mut self.cursor),
+            IxdtfOptions::Time => time::parse_annotated_time_record(&mut self.cursor),
+        }
```
Comment (sffc) [inline utils/ixdtf/src/parsers/mod.rs:92]: Thought: With big match statements like this, it is better for DCE if we instead had four functions like `parse`, `parse_year_month`, `parse_month_day`, and `parse_time`. Then we wouldn't need the enum.

File: utils/ixdtf/src/parsers/records.rs [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+pub enum TimeZoneRecord<'a> {
+    /// TimeZoneIANAName
+    Name(&'a str),
+    /// TimeZoneOffset
+    Offset(UTCOffsetRecord),
+}
+
+/// The parsed sign value, representing whether its struct is positive or negative.
+#[repr(i8)]
+#[allow(clippy::exhaustive_enums)] // Sign can only be positive or negative.
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:87]: Question: there are often 3 types of signs, known as `Signum`, which are "explicit minus", "explicit plus", and "no sign", as in "-1", "+1" and "1". Does IXDTF make a distinction?

File: utils/ixdtf/src/parsers/records.rs [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+#[cfg(feature = "duration")]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DurationFraction {
+    /// The fraction value applied to an hour.
+    Hours(u64),
+    /// The fraction value applied to the minutes field.
+    Minutes(u64),
+    /// The fraction value applied to the seconds field.
+    Seconds(u32),
+}
```
Comment (sffc) [inline utils/ixdtf/src/parsers/records.rs:157]: Thought: another way to model this, which may or may not be better, would be

```rust
pub enum TimeDuration {
    None,
    Hours {
        hours: u32,
        fraction: u64,
    },
    Minutes {
        hours: u32,
        minutes: u32,
        fraction: u64,
    },
    Seconds {
        hours: u32,
        minutes: u32,
        seconds: u32,
        fraction: u32,
    }
}
```

This more properly encodes the invariant that if hours have fractions, then minutes and seconds don't exist, for example.

**Review by sffc — APPROVED** [review]
Praise: This is the best version yet! Thanks for all the work you've been putting into this. All my comments can be done later or filed as a follow-up issue. I think we can land the first version now!

File: utils/ixdtf/src/parsers/annotations.rs [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+                calendar_crit = kv.critical;
+                continue;
+            }
+
+            if calendar_crit || kv.critical {
+                return Err(ParserError::CriticalDuplicateCalendar);
+            }
+        }
+
+        annotations.push(kv);
```
[the remark this answers is not in the document]
Comment (nekevss) [reply utils/ixdtf/src/parsers/annotations.rs:77]: That's probably true. I almost switched it (like with `TimeZoneAnnotation`), but given that `u-ca` is a registered key, I was viewing it as the parser being the primary consumer, so it was fine to leave it without the critical flag included. After looking back at the spec draft, I'm conflicted but leaning a bit more towards include the annotation.

For clarity on the point in the specification I'm thinking about mostly:

```txt
   An application that encounters duplicate use of a suffix key in
   elective suffixes and does not want to perform additional processing
   on this inconsistency MUST choose the first suffix that has that key,
   i.e.,

   2022-07-08T00:14:07Z[u-ca=chinese][u-ca=japanese]
   2022-07-08T00:14:07Z[u-ca=chinese]

```

All that considered and given the current approach to this parser, I think it makes sense to set calendar to the `Annotation` for the first calendar and all the rest are moved to `annotations` to be handled by the user, i.e. 1 & 3, but default to the first calendar over selecting calendars

Secondary Idea:
Would it make sense to add an opt-in `strict` parsing option to `IxdtfParser` for calendar / critical flag handling? When `strict` the parser would through an error on any strict calendar ambiguity or unknown critical annotations, and when not strict the calendar ambiguity and unknown critical annotations are provided to the user to handle?

File: utils/ixdtf/src/parsers/records.rs (hunk 27 above)
Comment (nekevss) [reply utils/ixdtf/src/parsers/records.rs:87]: IXDTF does not to my knowledge and from what I can find. RFC3339 does layout that the offset must be "+" or "-". So Duration is the only one that can be "no sign" as it is considered optional.
